### ISSUE TITLE
Storage Item Serialization and Validation

### ISSUE THREAD

**Issue by aklinker1** [wxt-dev/wxt#1173]
### Feature Request

When defineing storage items, getting values, or setting values, we should support serialization options for converting storage to and from a specific data type:

```ts
const enabledSites = storage.defineItem<Set<string>, {}. string[]>("enabledSites", {
  schema: {
    // Convert value to storage value
    serialize: (set) => [...set],
    // Convert storage value to usable type
    deserialize: (array) => new Set(array),
  },
});

const set = await enabledSites.getValue();
set.has("google.com")
```

A natural extension to this is to support various other popular data serialization and validation libraries, like Zod and TypeBox.

```ts
import { z } from 'zod';

const themePreference: WxtStorageItem<"light" | "dark" | "system"> = storage.defineItem("themePreference", {
  schema: z.enum(["light", "dark", "system"]),
});
``` 

```ts
import { Type } from '@sinclair/typebox';

const themePreference: WxtStorageItem<"light" | "dark" | "system"> = storage.defineItem("themePreference", {
  schema: Type.Union([Type.Literal("light"), Type.Literal("dark"), Type.Literal("system")]),
});
``` 

Should probably also support a `schema` option for `storage.getItem` as well.

#### Is your feature request related to a bug?

https://github.com/wxt-dev/wxt/discussions/1123 

### What are the alternatives?

Write a wrapper around `storage.defineItem` yourself.

### Additional context

N/A

**Comment by Timeraa** [comment]
Would be nice with arktype

**Comment by aklinker1** [comment]
> Would be nice with arktype

Hadn't heard of that one before... I'll give it a go sometime, looks really cool! I see you're a sponsor lol.

Should be easy enough to support, just a function, or we can use the assert function that integrates with tRPC

```ts
type StorageItemSchema<TValue, TStored> =
  // Manual serialization functions
  | { serialize(value: TValue): TStored, deserialize(stored: TStored): TValue }
  // Zod
  | { parse(stored: TStored): TValue }
  // Arktypes & Typebox
  | (stored: TStored) => TValue
```

**Comment by aklinker1** [comment]
Will also need to decide what happens when validation fails... Do we do nothing? Return the default value? Overwrite the bad value or leave it as-is? How do we integrate a logger for error reporters to hook into? Some questions to answer...

**Comment by Timeraa** [comment]
Not quite sure how you'd design the api around it but might be nice to have the option to decide what happens, like either hook into an event and if it's not hooked into we reset it or something?

**Comment by Timeraa** [comment]
> > Would be nice with arktype
> 
> Hadn't heard of that one before... I'll give it a go sometime, looks really cool! I see you're a sponsor 😉
> 
> Should be easy enough to support, just a function, or if something goes terribly wrong, we can always use the assert function they show integrating with tRPC.
> 
> ```ts
> type StorageItemSchema<TValue, TStored> =
>   // Manual serialization functions
>   | { serialize(value: TValue): TStored, deserialize(stored: TStored): TValue }
>   // Zod
>   | { parse(stored: TStored): TValue }
>   // Arktypes & Typebox
>   | (stored: TStored) => TValue
> ```

I love arktype, it's super fast and gives the best usable types in the editor imo haha

**Comment by aklinker1** [comment]
One more addition: an alternative to Plasmo's secure storage. I don't think calling it "secure" is the right move, in reality, it's just 2-way encoded with a "secret" key that's shipped with the extension. So we could easily add an equivalent API:

```ts
const encodeAtRest = createEncode("key");

export const example = defineStorageItem("local:example", {
  schema: encodeAtRest(z.String())
});
```

Doesn't need to be apart of the initial PR, and that's just an idea of what the API might look like. Could be a separate field unrelated to `schema` as well.

**Comment by aiktb** [comment]
A limitation in chrome.storage is that proxy arrays are converted to regular objects when written, making all array methods unavailable when read.
If WXT storage had serialization capabilities, this limitation could be easily overcome.
This is crucial for Vue Devs.

**Comment by Iceman8911** [comment]
Is there any progress on this?

**Comment by silverfish2525** [comment]
@aklinker1 why not use Standard Schema for the parse branch? Zod/Valibot/ArkType/Effect all conform, so any of them just works. Keep the codec branch separate though, cause  `Set`/`Map`/`Date` need round-trip and Standard Schema's `validate` is one-way only. Same for the function branch, since TypeBox and custom parsers aren't `parse`-shaped.

   ```ts
   type StorageItemSchema<TValue, TStored = unknown> =
     | { serialize(v: TValue): TStored; deserialize(s: TStored): TValue }
     | StandardSchemaV1<unknown, TValue>
     | ((stored: unknown) => TValue);
 ```

 On validation failure, configurable, default throw:

 ```ts
   onValidationError?: 'throw' | 'fallback' | 'reset' | ((issues, raw) => TValue);
 ```

 Reads only though. If setValue gets bad data we should throw. Silently swapping in a fallback might hide bugs.

 For the third question, we should migrate then validate. Otherwise old data that needs migrating just fails validation before migrate can run.

 We can have a new defineItem overload, plus the schema has to reach get, set, migrate and watch.

 On the adjacent stuff: encoder is basically the codec branch applied at rest, so no new API once this lands. getItems/setItems is mostly a typing job, though the schema does need to thread through those return types. And metaSchema is the same
 pattern on .getMeta()/.setMeta(), easy to add once the value side works.

 Does this direction look correct to you? Should I open a PR?

**Comment by aklinker1** [comment]
Thinking about this more, there's really two distinct things here: serialization and validation. I know my original proposed API combined them under one option, but is that really the right way? For example, storing a date... What would `storage.defineItem` look like? A standard schema definition isn't enough to serialize the date into a string, but, with `z.coerce.date()`, it would be enough to deserialize it.

So I think we need to allow passing both serialization functions and a standard schema object (yes, I agree, we should use it).

So instead of `StorageItemSchema` as you've described, maybe we go for something more like this:

```ts
export const installDateStorage = storage.defineItem('local:install-date', {
  schema: {
    serialize: (date) => date.toString(),
    deserialize: (str) => new Date(str),
    validate: z.date(),
    // onValidationError: ... what you proposed, I like that
  },
})
```

Or fully separate the two flows, like this:

```ts
export const installDateStorage = storage.defineItem('local:install-date', {
  transform: {
    serialize: (date) => date.toString(),
    deserialize: (str) => new Date(str),
  },
  validate: {
    schema: z.date(),
    // onValidationError: ...
  },
})
```

Or this:

```ts
export const installDateStorage = storage.defineItem('local:install-date', {
  transform: {
    serialize: (date) => date.toString(),
    deserialize: (str) => new Date(str),
  },
  schema: z.date(),
  // onValidationError: ...
})
```

Basically, I'm just saying we probably don't want to cram too much functionality into a single option. These two things seem like they could be implemented independent of each other. I think I'd prefer the final API from above, with `schema` on the root of the options object and serialization as a nested object.

I'm also fine with using different names for all these options. [Vueuse uses "serializer" and "read" and "write"](https://vueuse.org/core/useStorage/#custom-serialization) instead of "transform", "serialize", and "deserialize". That's probably better then what I proposed, but I would be open for any other proposed naming.

---

So to directly answer your questions:

1. Yes to standard schema. I believe that came about after this issue was created, or it hadn't seen as much adoption as there is today.
2. I don't like your `StorageItemSchema` type and my originally proposed API - it doesn't allow for both serialization and validation, see above for alternative proposals - I would like your (and anyone else's) feedback on it.
3. I like your proposal for handling validation errors - the `onValidationError` sounds good, and yes, I agree it should only be followed when a value is read from storage. Setting an invalid value should always throw.
4. Yes, migrate then validate.
5. Let's leave out metadata for now, if it's something we want to validate/serialize, let's create a separate issue for it, or at least do it in a separate PR. This one will be big enough as-is.

**Comment by silverfish2525** [comment]
yeah the Date example does it, they're two things.

   going with your third shape, schema at root, serializer nested. simplest.

   one bit i want us to agree on since your example implies it but doesnt say: `validate: z.date()` puts the schema on the deserialized runtime value, not the raw wire form. so
 read is `raw → migrate → deserialize → validate → T`, write is `T → validate → serialize → raw`. calling it out cause otherwise someone shows up with `schema:
 z.string().datetime()` for a Date and asks why the types dont line up.

   that ordering also makes `serializer.deserialize` optional, coerce schemas cover it. Date works either

   ```ts
   serializer: { serialize: d => d.toISOString(), deserialize: s => new Date(s) },
   schema: z.date(),
   ```

   or

   ```ts
   serializer: { serialize: d => d.toISOString() },
   schema: z.coerce.date(),
   ```

   docs would probably steer people to coerce for common types.

   naming, yeah VueUse verbatim, `serializer: { read, write }`.

   for TypeBox and hand-rolled parsers i looked at what the ecosystem does. nothing ships per-library adapter packages — @hono and @kori validators are framework-side, not
 TypeBox→StandardSchema wrappers. its just a small adapter helper that wraps a parse function into a Standard Schema object. fits the defineX pattern:

   ```ts
   schema: defineSchema<MyType>((v) => Value.Parse(myTypeboxType, v))
   ```

   TypeBox, io-ts, custom validators all use the same adapter, `schema` stays a single type not a union.

   couple things your reply doesnt touch, id like to nail them down since the design differs if we skip them:

   - `watch(cb)`: schema is set, invalid value comes in from another context, what does cb see? id say run the same read pipeline for watch and `onValidationError` fires there too.
 otherwise a schema-guarded item still calls its watchers with unvalidated data.
   - `getValueOrInit`: init value should validate before write, else a broken init sits at rest till someone reads it.

   scope id have in mind: new `defineItem` overload with `schema` + `serializer`, thread through get/set/watch/getValueOrInit + migrations, `defineSchema` helper, tests per schema
 branch and per `onValidationError` mode. metadata + encoder + defineMap stay separate as agreed.

**Comment by bn-dev-042** [comment]
If anyone is interested, here is my implementation of _store with schema_:

https://github.com/botnadzor/extension/blob/b820875a5089ebb430c1e650a0a7f56b3b7fadc6/src/entrypoints/background/%40service-helpers/store-with-schema.ts#L1

Usage

```ts

const myConfigStore = defineStoreWithSchema("sync:my-config", myConfigSchema);
```
